src/support/qsort.c

```diff
@@ -0,0 +1,263 @@
+            /*
+             * For larger arrays, do more work to find a good median - here, it's a "pseudo-median"
+             * of nine samples.
+             */
+            if (nmemb > WT_QSORT_LARGE_THRESHOLD) {
+                size_t median_dist = (nmemb / 8) * elem_sz;
+
+                lo_pseudomedian = __med3(lo_pseudomedian, lo_pseudomedian + median_dist,
+                  lo_pseudomedian + 2 * median_dist, cmp, context);
+                pseudomedian = __med3(pseudomedian - median_dist, pseudomedian,
+                  pseudomedian + median_dist, cmp, context);
+                hi_pseudomedian = __med3(hi_pseudomedian - 2 * median_dist,
+                  hi_pseudomedian - median_dist, hi_pseudomedian, cmp, context);
+            }
+            pseudomedian = __med3(lo_pseudomedian, pseudomedian, hi_pseudomedian, cmp, context);
+        }
+
+        __swap_bytes(a, pseudomedian, elem_sz);
+        lowest_lt_median = lo_unknown = a + elem_sz;
+        hi_unknown = highest_gt_median = a + (nmemb - 1) * elem_sz;
+        for (;;) {
+            while (lo_unknown <= hi_unknown && (cmp_result = CMP(lo_unknown, a, context)) <= 0) {
+                if (cmp_result == 0) {
+                    swapped = true;
+                    __swap_bytes(lowest_lt_median, lo_unknown, elem_sz);
+                    lowest_lt_median += elem_sz;
+                }
+                lo_unknown += elem_sz;
+            }
+
+            while (lo_unknown <= hi_unknown && (cmp_result = CMP(hi_unknown, a, context)) >= 0) {
+                if (cmp_result == 0) {
+                    swapped = true;
+                    __swap_bytes(hi_unknown, highest_gt_median, elem_sz);
+                    highest_gt_median -= elem_sz;
+                }
+                hi_unknown -= elem_sz;
+            }
+
+            if (lo_unknown > hi_unknown)
+                break;
+
+            __swap_bytes(lo_unknown, hi_unknown, elem_sz);
+            swapped = true;
+            lo_unknown += elem_sz;
+            hi_unknown -= elem_sz;
+        }
+
+        if (!swapped) {
+            /* Switch to insertion sort. */
+            for (hi_unknown = a + elem_sz; hi_unknown < a + nmemb * elem_sz; hi_unknown += elem_sz)
+                for (lo_unknown = hi_unknown;
+                     lo_unknown > a && CMP(lo_unknown - elem_sz, lo_unknown, context) > 0;
+                     lo_unknown -= elem_sz)
+                    __swap_bytes(lo_unknown, lo_unknown - elem_sz, elem_sz);
+            return;
+        }
+
+        hi_pseudomedian = a + nmemb * elem_sz;
+
+        WT_ASSERT(NULL, lowest_lt_median >= a);
+        WT_ASSERT(NULL, lo_unknown >= lowest_lt_median);
+        WT_ASSERT(NULL, highest_gt_median >= hi_unknown);
+        WT_ASSERT(NULL, hi_pseudomedian >= (highest_gt_median + elem_sz));
+
+        /* Swap our pivot back into the correct place. */
+        lhs_unsorted =
+          WT_MIN((size_t)(lowest_lt_median - a), (size_t)(lo_unknown - lowest_lt_median));
```

Comment:
Can we note the reason for the casts here? I think it would be better to call them out.

Reply:
Good spot - thanks, fixed in 35eedc7.